### ISSUE TITLE
Videos do not play properly. No audio.

### ISSUE THREAD

**Issue by projectthobunny** [ankidroid/Anki-Android#5259]
###### Reproduction Steps

1.  Create a subs2srs deck using a YouTube video
2.  Try to review cards

###### Expected Result

Videos should play with full audio.

###### Actual Result

Most videos have no audio, while some have partial audio. I suspect this has to do with missing keyframes.

###### Debug info

AnkiDroid Version = 2.8.4
Android Version = 9
ACRA UUID = d4e90276-227e-4efb-a22b-448c58d91243

###### Research
*Enter an [ x ] character to confirm the points below:*

[ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x ] I have searched for similar existing issues here and on the user forum

#######

I am using a Pixel, so this issue is on stock Android. These video files play perfectly on desktop and iPhone.

I suspect subs2srs does not re-encode each individual clip, and simply cuts them at the specified times.
I also suspect the default android video player does not properly support videos that are missing keyframes. 

I would guess there are five solutions to this problem.
1. AnkiDroid bundles a simple video player, like Anki does
1. AnkiDroid allows the user to select a default video player
1. AnkiDroid implements a workaround that embeds the videos into the card (If I manually embed the videos they work)
1. subs2srs changes their software
1. I manually re-encode my entire video collection

Thank you for looking into this.

**Comment by timrae** [comment]
> I would guess there are five solutions to this problem.

1. I'm opposed to this, as it will increase APK size and cause extra work for us
1. Possible but you'd most likely have to implement it yourself in a clean way that we agreed upon
1. I'm basically opposed to this as well, however if you convinced @dae to do this by default (on desktop, web, and iOS), it would be a different story, and it could be the cleanest way to get media working on all platforms in the long-term. In the meantime you could make a plugin for Anki desktop that makes it easy for you to embed videos from the note editor when you're creating new cards.
1. You should definitely request them to add support for Android and not just iPhone. Again, you may have to implement this yourself though, as it's probably not a priority for them.
1. This is probably the fastest and easiest way, though it doesn't solve the problem in the long-term. You can make a simple shell script to avoid doing it manually for each file.

**Comment by projectthobunny** [comment]
3. ...if you convinced @dae to do this by default (on desktop, web, and iOS)...

You know, now that I think about it, he may already be doing this with AnkiMobile. It certainly looks embedded, but I don't know exactly what is going on behind the scenes.

![ankimobile_video](https://user-images.githubusercontent.com/47655181/53859874-7eaf6380-4022-11e9-969f-4f539ab1caf7.png)

**Comment by projectthobunny** [comment]
### Couldn't fix it through re-encoding
So I tried for a couple hours to re-encode my videos but never reached a satisfactory resolution. I was able to get the videos to consistently play _some_ audio, but the videos also consistently started playing at ~1s. Considering how short the clips are, that is a deal breaker for me and anyone else using SRS sized clips.

Of course, someone with deeper knowledge of video encoding might be able to find a solution. I am certainly not saying it is impossible to fix on the encoding end. But a lack of response from the subs2srs developer, as well as the failure of simple re-encoding to fix the problem, leads back towards updating AnkiDroid as the solution.

### Embed the videos?
I suspect solution 3 (embed videos) would be by far the easiest to implement, though solution 2 (allow select default player) _could_ work, 

Anki already detects video/audio files and treats them differently, right? Instead of having a `{{video clip}}` trigger opening the player, why not just trigger replacing the `{{video clip}}` with a little html code block using the `<video>` tag? Some simple string manipulation would remove the '[sound:' and ']', and that would be it.

`{{video clip}}` => `[sound:VIDEO.mp4]` =>

```
<video width="320" height="240" controls>
  <source src="VIDEO.mp4" type="video/mp4">
</video>
```

At the very least it would be great to have this available in the settings. Although video functionality currently "exists", in practice it doesn't work in one of the most common use case scenarios.

Is there any reason this solution wouldn't work? Obviously if I had the skills I would implement it myself, but looking from the outside in it seems like it could be relatively painless to implement.

**Comment by timrae** [comment]
> Is there any reason this solution wouldn't work?

It should work fine, but probably only on Android 4.4+. We will eventually drop support for older Android versions, so that should be fine, but probably not until after 2.9 is released (no idea on when this will be tbh).

> looking from the outside in it seems like it could be relatively painless to implement

The chance of an existing maintainer picking this up are basically zero, for the simple reason that AFAIK none of us are using video. So you'll either need to implement this yourself (which is probably not as hard as you think, see the [development wiki](https://github.com/ankidroid/Anki-Android/wiki/Development-Guide) for getting started), pay someone to do it for you, or wait on the off chance that someone fixes it for you. If it works well for you, then please submit a pull request and we can hopefully merge it in the medium term future.

**Comment by projectthobunny** [comment]
Clearly the solution here is to convince the maintainers to start using videos.

I really appreciate you taking the time to respond to my questions. I don't currently speak java, but I'll give it a shot when I find time. The development wiki makes it look doable.

Thanks again.

**Comment by kelciour** [comment]
As an alternative to subs2srs you may try to use [movies2anki](https://ankiweb.net/shared/info/939347702) as an Anki 2.0 add-on or a standalone app.
As far as I know, Anki 2.1 isn't built with proprietary media codecs support ([Qt WebEngine Features](https://doc.qt.io/qt-5/qtwebengine-features.html#audio-and-video-codecs)) and `MP4` won't work with Anki 2.1 (though it may or may not work with AnkiDroid or AnkiMobile). 
There may be some issues with `<video>` tag, but as far as I know `WEBM` ([VP8](https://trac.ffmpeg.org/wiki/Encode/VP8) or [VP9](https://trac.ffmpeg.org/wiki/Encode/VP9)) should work on AnkiMobile, AnkiDroid and Anki 2.1. Please see this card template - https://yadi.sk/d/teEDwpTUI5mxcQ
If there's no audio, it may help to reencode audio with `libvorbis` instead of `libopus`.
